### ISSUE TITLE
Update to futures 0.3

### ISSUE THREAD

**Issue by dlrobertson** [servo/ipc-channel#219]
## Summary
Move from futures (0.1) to futures-preview (0.3).

## Details

Update the `futures` crate for `futures-preview` and update the implementation of `Stream` for `IpcBytesReceiver` and `IpcReceiver` to reflect this change. Between futures 0.1 and futures 0.3 the API was changed to include notification. In order to move to futures 0.3 `Stream::poll_next` needs to be able to use a `LocalWaker` to notify the task that we need to `poll` for new data.

**Comment by asajeffrey** [comment]
It would be really really nice if we could do this. @antrik do you know if we can rely on the native libraries giving us some way to do this? We'd need some way of attaching a callback to an ipc channel which is called whenever new data arrives. If not, is there some way to select on multiple IPC channels, so we'd only have to dedicate one thread to waking? Otherwise, we'd need to dedicate one thread per IPC channel, which is quite expensive.

**Comment by antrik** [comment]
@asajeffrey sorry, I'm not familiar with the futures system, and thus don't quite understand what is needed here... Could you try to elaborate?

**Comment by asajeffrey** [comment]
@antrik We'd need the ability to register a callback, which is called whenever new data arrives on the ipc channel.

It would be fine to dedicate a thread-per-process to this, but probably not a thread-per-ipc-channel. Unfortunately this would require being able to select on ipc channels, which I'm not sure we can do.

**Comment by antrik** [comment]
Just to be clear: when a channel wrapped in a feature gets data, we simply need to make sure a callback is run in some sort of dispatcher thread?

Does it have to be a pure notification, only informing us of readiness, like the kernel `poll()`/`select()` calls? Or is it fine if it reads and returns the actual data that arrived? Because if it's the latter, I'm confused why you are asking, considering that `ipc-channel` always has had this... Am I missing something?

(I wonder about cancellation requirements, though: I guess it might be a problem that with the current interface, once a receiver has been added to a set, there is no way to extract or remove it? We could extend the internal API to allow for that, if necessary -- or maybe even the public one, if it's deemed generally useful...)

**Comment by asajeffrey** [comment]
In https://github.com/rust-lang-nursery/futures-rs/issues/1466#issuecomment-468712967, @seanmonstar points out that the current implementation has the same problem, the only thing that's changed is that the waker is a parameter to `poll` rather than being a thread-local variable.

**Comment by asajeffrey** [comment]
@antrik yes, we just want a callback to be called when data arrives. It's fine if the notification also reads the data. Ah, `IpcReceiverSet` at least means we only need one thread per process to implement the notification, that's not so bad.

The interface for `IpcReceiverSet` gives access to the serialized bytes, so the caller is expected to do the deserialization? Does the caller get any guarantees about how the data arriving form the `IpcReceiverSet` lines up with the serialization?

**Comment by asajeffrey** [comment]
Ah, did more digging into the `OpaqueIpcMessage` type, looks like it does exactly what we need in terms of getting one message from the select corresponding to one message from the sender. OK, I can see how to do the implementation now, with one thread per process dedicated to waking up ipc channels.

**Comment by asajeffrey** [comment]
Hmm, some versioning issues...

- Do we want to continue to support futures 0.1 as well as 0.3?
- futures 0.3 requires nightly (for the `Pin` api), but futures 0.1 compiles on stable, is this a problem?

**Comment by asajeffrey** [comment]
OK, I wrote a first shot at this. This is an implementation of a router that forwards IPC channels to future channels.

I should work out whether we can share code between this and the router to mspc channels.

Rather annoyingly, I ended up needing to spin up two threads, in order to route a wake-up message from an mspc channel to an IPC channel, since IPC channels aren't `Sync`, so can't be used in `lazy_static!`.

Current draft at https://github.com/asajeffrey/ipc-channel/tree/futures-0.3
